### ISSUE TITLE
I tried DeepNextFace myself, but landmark loss decreases slowly.

### ISSUE THREAD

**Issue by anewusername77** [abdallahdib/NextFace#37]
Hey there, thanks for your impressive work!
I tried Nextface first and now implemented DeepNextFace by myself, using default resnet152 in pytorch(pretrained on ImageNet), but during step 1, landmark loss decreased in a quite slow way(landmark loss is the same as NextFace), ended up with about 3000. 
I think it may related to slow fitting of focal and camera position params. Dosen't find more details about the training strategy in your paper including appendix, can you please help?

**Comment by abdallahdib** [comment]
Hi, 
are u using only the landmark loss for training? what are ur learning rates? does the loss converges? can u show some intermediate output? please provide more details to see how i could help. 3000 as a loss looks weird to me.

**Comment by anewusername77** [comment]
Thanks a lot for your reply, I'll write as much details as I can think of.
Dataset I use is `celebA 256x256`, 30000 images in total.
landmarks are computed by `face_lignment mediapipe`, outputs of Resnet152 are:
```
shape_coeff = output_p[:, :80]
exp_coeff = output_p[:, 80:155]
albedo_coeff = output_p[:, 155:235]

sh_coeffs = output_p[:, 235:478].reshape(-1, self.cfg.SH_bands * self.cfg.SH_bands, 3)
# camera coeff
focals = output_p[:, 478]
rotation = output_p[:, 479:482]
translation = output_p[:, 482:485]
```
At first stage(I'm wonder if there is any pre-training stage), I trained with regStatModel(of expression, shape and albedo), landmark loss and photo loss, weights of them are the same as your paper said:
![image](https://user-images.githubusercontent.com/23047054/189787789-6c596105-38b9-4a3c-bab7-97138b7070df.png)
```python
landmarkloss = landmarkLoss(self.BFM.landmarksAssociation, cam_vertices, landmarks, focals, self.cam_center)
reg = self.cfg.wi * (regStatModel(shape_coeff, self.BFM.shape_pca_var) + \
                             regStatModel(exp_coeff, self.BFM.expression_pca_var)) + \
         self.cfg.wc * regStatModel(albedo_coeff, self.BFM.diffuse_albedo_pca_var)
ptloss = 1000 * photoLoss(smoothedImage, torch.pow(x.permute(0, 2, 3, 1), 2.2), mask)
loss = landmarkloss + reg + ptloss
```

learning rate decays per iteration from `1e-4(cosine decay)`

when running like this, I can't get any output from rendering, all black(camera can't see the object at all), landmark loss ends up with 3000+. Thus, I added two L1 loss for `translation, rotation` of network and that of function `initCameraPos`, goes like this:
```
1000*ad_loss1(rotation, rots) + ad_loss2(translation, trans)
```
`trans and rots` are used for transVertices, not updated during back propagation,
**after training 35 epoch with batch_size 8, landmark loss ends up with `10~15`, 1000*photo loss ends up with `120~140`, L1 losses end up with `9~15`, reg ends up with `0.3`**
intermediate output is like（8 sample）:
![image](https://user-images.githubusercontent.com/23047054/189791377-866efb9b-ca00-4d9e-933b-9d15afcfc4a2.png)

If I give `shape reg` a large penalty, 1000 for example, it will ends up with a mean face.

**Comment by abdallahdib** [comment]
Hey, u should take note that in the paper we were using basel 2009 where here u are using basel 2017. ur Regularizations terms need to be revisited and adjusted. i suggest working on small dataset of hundred or thousands of images to do the hyper parameters tuning. 

Plz also increase the landmarks reg weight to be equal to the photo loss. In the paper the photo loss was normalized. 
Also the landmarks loss term is important especially at the beginning of the training to avoid trivial solution where the face is pushed too far so that the photo loss is equal to zero  (u can relax the landmarks weight  after few epoches). I suggest starting with large landmarks weights and progressively relax it. 
Also can u verify on ur side that the gradients are there for the rotation and translation part? i think there is a bug in NextFace that clear the gradients on rotation and translation and i need to fix it in a future push. So plz verify ur gradients

I hope that this help u. dont hesitate to ask if u need more help.

**Comment by anewusername77** [comment]
Hi, thanks for your patience and help, I'll try small dataset and large landmark loss weight with deleting regularization of translation and rotation.

To be sure I understand your suggestion: 
1. `Regularizations terms need to be revisited and adjusted`, meaning in DeepNextFace paper 
![image](https://user-images.githubusercontent.com/23047054/190884150-df915009-ec89-4a7c-bf10-cf849398798a.png)
only included shape and albedo reg, but I should change relugarizations term to as they are in NextFace Code `runstep 2`  , since using BFM2017. 
2. `increase the landmarks reg weight to be equal to the photo loss`, meaning set penalty of landmark loss equal to photoloss? 1000 for example?
3. learning rate of mine is okay to train encoder and don't need to train as `runstep1` in NextFace at the beginning(just using landmark loss, reg and photoloss togther when training encoder is fine)
5. yes, I'm sure rotation and translation from network `requires_grad=True`, and are updating during traning, while rots and trans from `initCameraPos`  `requires_grad=False`. I also found the bug as you said in NextFace, and I think **swap the position of  self.enableGrad() and self.initCameraPos() in optimizer.py(line 136, 137)** should help to make graients be there for them, results of which are slightly better than that of your original code. So I think rots and trans of `iniCameraPos` are already good enough to give a proper render result in NextFace.

One more thing, is it possible for you to give me some reference results after first starge training (training only encoder)?
![image](https://user-images.githubusercontent.com/23047054/190883996-c8085a0c-4b49-473a-8531-bd944c14646e.png)
like value of landmarkloss, photoloss, exp reg, shape reg and albedo reg.

**Comment by anewusername77** [comment]
I tried `resnet152` with `lr=1e-4` and loss as:
``` python
lmloss = 1000 * landmarkLoss(self.BFM.landmarksAssociation, cam_vertices, landmarks, focals, self.cam_center)
# same reg as NextFace code runstep2
reg = 0.0001 * sh_coeffs.pow(2).mean()
        + 0.001 * regStatModel(albedo_coeffs, self.BFM.diffuse_albedo_pca_var)
        + 0.001 * regStatModel(shape_coeffs, self.BFM.shape_pca,_var)
        + 0.001 * regStatModel(exp_coeffs, self.BFM.expression_pca_var)
photoloss = photoLoss(smoothedImage, torch.pow(x.permute(0, 2, 3, 1),2.2), mask)
loss = lmloss + reg + photoloss
```
plus `focals`, `rotations` and `translations` are from network, deleting `initCameraPos`, but this time, focal is always negative, `lmloss` keeps around 6,000,000. Model can't converge.

It seems that my resnet152 pretrained from ImageNet can't converge when **using translation and rotation from the net for `transVertices`** without the help of rots and trans from `initCameraPos`, also focals can't be trained to a proper value.

Is there anything I missed during training.....would you plz tell me about ur training strategy(are they same as that in your DeepNextFace Paper's appendix? especially learning rate and loss...)

sorry for asking so many questions, best regards.

**Comment by anewusername77** [comment]
still waiting for your reply (T T)...

**Comment by rlczddl** [comment]
@anewusername77 I have encountered problems like yours recently.(如果可以的话,可以留个联系方式讨论下)

**Comment by anewusername77** [comment]
这是我的邮箱-------- 原始邮件 --------发件人： rlczddl ***@***.***>日期： 2022年9月27日周二 晚上8:16收件人： abdallahdib/NextFace ***@***.***>抄送： 一口酥 ***@***.***>, Mention ***@***.***>主    题： Re: [abdallahdib/NextFace] I tried DeepNextFace myself, but landmark loss decreases slowly. (Issue #37)

@anewusername77 I have encountered problems like yours recently.(如果可以的话,可以留个联系方式讨论下)

**Comment by anewusername77** [comment]
有啥问题邮件沟通吧

**Comment by abdallahdib** [comment]
just a simple initialization of the layer (set weights to 0 and bias to a constant is enough)

**Comment by anewusername77** [comment]
get it!

**Comment by anewusername77** [comment]
I'm back. 
I tried, a good thing is landmarkloss decreased faster than before, but trans and rots seemed weird, the face was also twisted cuz shape and expression reg were big.  here;s the intermediate result.
```
lmloss:100.21551602544034       
reg(before multiplying 0.001): shape: 40.70402145385742, express: 46.430816650390625
focal:34.22245788574219 
trans: tensor([ 20.2277, 107.4193, 110.7016], device='cuda:0', grad_fn=<SelectBackward0>)       
rots:tensor([ 2.3135,  0.2383, -0.1689], device='cuda:0', grad_fn=<SelectBackward0>)
```

**Comment by abdallahdib** [comment]
plz can u share visual results and code snippest

**Comment by anewusername77** [comment]
您好，请问您现在复现得怎么样了？-------- 原始邮件 --------发件人： rlczddl ***@***.***>日期： 2022年9月27日周二 晚上9:11收件人： abdallahdib/NextFace ***@***.***>抄送： 一口酥 ***@***.***>, Mention ***@***.***>主    题： Re: [abdallahdib/NextFace] I tried DeepNextFace myself, but landmark loss decreases slowly. (Issue #37)

没看到 你加我吧微信:caitong94















---Original---



From: ***@***.***&gt;



Date: Tue, Sep 27, 2022 20:57 PM



To: ***@***.***&gt;;



Cc: ***@***.******@***.***&gt;;



Subject: Re: [abdallahdib/NextFace] I tried DeepNextFace myself, but landmarkloss decreases slowly. (Issue #37)



















 这是我的邮箱-------- 原始邮件 --------发件人： rlczddl ***@***.***&gt;日期： 2022年9月27日周二 晚上8:16收件人： abdallahdib/NextFace ***@***.***&gt;抄送： 一口酥 ***@***.***&gt;, Mention ***@***.***&gt;主    题： Re: [abdallahdib/NextFace] I tried DeepNextFace myself, but landmark loss decreases slowly. (Issue #37) 



 @anewusername77 I have encountered problems like yours recently.(如果可以的话,可以留个联系方式讨论下)

**Comment by rlczddl** [comment]
我没有复现这个项目，我这边是想把depth加进来和color一起优化。目前来看，需要一个好的rt作为初始值，pnp在我这边有时的解出来的rt有时候有问题，导致网络优化其它参数出现问题。



---Original---
From: ***@***.***&gt;
Date: Thu, Oct 13, 2022 12:58 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [abdallahdib/NextFace] I tried DeepNextFace myself, but landmarkloss decreases slowly. (Issue #37)




 您好，请问您现在复现得怎么样了？-------- 原始邮件 --------发件人： rlczddl ***@***.***&gt;日期： 2022年9月27日周二 晚上9:11收件人： abdallahdib/NextFace ***@***.***&gt;抄送： 一口酥 ***@***.***&gt;, Mention ***@***.***&gt;主    题： Re: [abdallahdib/NextFace] I tried DeepNextFace myself, but landmark loss decreases slowly. (Issue #37) 
 没看到 你加我吧微信:caitong94 
 
 
 
 
 
 
 
 ---Original--- 
 
 From: ***@***.***&amp;gt; 
 
 Date: Tue, Sep 27, 2022 20:57 PM 
 
 To: ***@***.***&amp;gt;; 
 
 Cc: ***@***.******@***.***&amp;gt;; 
 
 Subject: Re: [abdallahdib/NextFace] I tried DeepNextFace myself, but landmarkloss decreases slowly. (Issue #37) 
 
 
 
 
 
 
 
 
 
  这是我的邮箱-------- 原始邮件 --------发件人： rlczddl ***@***.***&amp;gt;日期： 2022年9月27日周二 晚上8:16收件人： abdallahdib/NextFace ***@***.***&amp;gt;抄送： 一口酥 ***@***.***&amp;gt;, Mention ***@***.***&amp;gt;主    题： Re: [abdallahdib/NextFace] I tried DeepNextFace myself, but landmark loss decreases slowly. (Issue #37)  
 
  @anewusername77 I have encountered problems like yours recently.(如果可以的话,可以留个联系方式讨论下)

**Comment by abdallahdib** [comment]
ur training code looks fine to me. Just initialize ur fully connected layers properly. 
Fr instance init the focal layer with a good values (3000). same for other semantic params. 
what is the order of magnitude of lmLoss compared to photoLoss?

**Comment by anewusername77** [comment]
do you mean weight of them? they are multiplied with the same weight as 1. ```loss = lmLoss + regLoss + photoLoss```
after training the first stage(train encoder only), landmark_loss ended up with around 100, shape_reg around 20. (face is strongly twisted)

**Comment by anewusername77** [comment]
hey there, I got a new question while trying decoder. It converges, but `diffuse texture loss` is much less than `spec texture loss`. For instance, at the very beginning of training decoder: `diff_txl:261.3556823730469,      spec_txl:171031.5625`. Is it reasonable？should I give an extra weight to spec_texure_loss?

btw, texture loss function I use is the same as Nextface, regularizers of which are the same as that in your paper: 
![image](https://user-images.githubusercontent.com/23047054/198182887-4d3a6f59-3391-42cd-9575-5eb87a238437.png)

**Comment by anewusername77** [comment]
sorry to bother you again.
landmark loss kept around 60 on celeba_align(2.4k imgs), even when i deleted rendering codes, meaning training with landmark loss only.
